### PR TITLE
Add Tag manager script

### PR DESCRIPTION
**Pull request by yaswant**
# PR Summary

Sci/Tech Reviewer: @r-sharp 
Code Reviewer: @t00sa 

Script to manage (add, delete, list) tags in a repository. 

## Code Quality Checklist

- [x] I have performed a self-review of my own code
- [x] My code follows the project's [style guidelines](https://metoffice.github.io/lfric_core/how_to_contribute/index.html#how-to-contribute-index)
- [x] Comments have been included that aid understanding and enhance the readability of the code
- [x] My changes generate no new warnings
- [x] All automated checks in the CI pipeline have completed successfully

## Testing

- [ ] This change has been tested appropriately (please describe)

## Security Considerations

- [ ] I have reviewed my changes for potential security issues
- [ ] Sensitive data is properly handled (if applicable)
- [ ] Authentication and authorisation are properly implemented (if applicable)

## AI Assistance and Attribution

- [ ] Some of the content of this change has been produced with the assistance of _Generative AI tool name_ (e.g., Met Office Github Copilot Enterprise, Github Copilot Personal, ChatGPT GPT-4, etc) and I have followed the [Simulation Systems AI policy](https://metoffice.github.io/simulation-systems/FurtherDetails/ai.html) (including attribution labels)

# Sci/Tech Review

- [ ] I understand this area of code and the changes being added
- [ ] The proposed changes correspond to the pull request description
- [ ] Documentation is sufficient (do documentation papers need updating)
- [ ] Sufficient testing has been completed

(_Please alert the code reviewer via a tag when you have approved the SR_)

# Code Review

- [ ] All dependencies have been resolved
- [ ] Related Issues have been properly linked and addressed
- [ ] Code quality standards have been met
- [ ] Tests are adequate and have passed
- [ ] Security considerations have been addressed
- [ ] Performance impact is acceptable

### CODE REVIEW

File: sbin/tagman [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    local target_ref
+    target_ref=$(gh api "repos/${REPO}/releases/tags/${REF}" --jq '.target_commitish')
+
+    # Resolve the target to full SHA
+    if resolved_sha=$(gh api "repos/${REPO}/commits/${target_ref}" --jq '.sha' 2>/dev/null); then
+      REF="$resolved_sha"
+      echo -e "${GRN}Resolved to commit: $REF${NC}"
+      return 0
+    else
+      echo -e "${RED}** Failed to resolve release target to commit SHA${NC}"
```
Comment (r-sharp) [inline sbin/tagman]:
```suggestion
      echo -e "${RED}** Failed to resolve release target to commit SHA${NONSENSICAL_COMMENT}"
```

File: sbin/tagman [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+  local url="https://github.com/${REPO}.git"
+  local msg="${MESSAGE:-"Tagging $TAG @ $REF"}"
+
+  if [[ "$DRY_RUN" == true ]]; then
+    echo -e "${YLW}[DRY RUN] Would create tag with the following details:${NC}"
+    echo -e "  Repository: $REPO"
+    echo -e "  Tag name:   $TAG"
+    echo -e "  Commit SHA: $REF"
+    echo -e "  Message:    $msg"
+    echo -e "${YLW}[DRY RUN] No changes made.${NC}"
```
Comment (r-sharp) [inline sbin/tagman]:
```suggestion
    echo -e "${YELLOW}[DRY RUN] No changes made.${NATURAL_CORD}"
```

File: sbin/tagman [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+delete_tag() {
+  [[ -z "$TAG" ]] && {
+    echo -e "${RED}** TAG is required for delete action.${NC}"
+    usage
+  }
+
+  verify_tag || {
+    echo -e "${RED}** Tag '$TAG' does not exist in repository '$REPO'.${NC}"
+    return 1
+  }
```
Comment (r-sharp) [inline sbin/tagman]:
```suggestion
  if ! verify_tag ; then
    echo -e "${RED}** Tag '$TAG' does not exist in repository '$REPO'.${NO_CHARGE}"
    return 1
  fi
```

File: sbin/tagman [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+    del|delete)
+      [[ $# -lt 1 ]] && usage
+      TAG="$1"
+      shift
+      ;;
+    ls|list)
+      # No arguments required
+      ;;
+    *)
+      echo -e "${RED}Unknown action: $ACTION${NC}"
```
Comment (r-sharp) [inline sbin/tagman]:
```suggestion
      echo -e "${RED}Unknown action: $ACTION${NORMAL_COUNTENANCE}"
```

**Comment by yaswant** [comment]
@r-sharp I appreciate you taking the time to review this thoroughly. Thanks for the suggestion re `jq` - I'll update the comment section to include it. However, I respectfully disagree with this assessment. I've replied to your individual comments/suggested changes as well.

Regarding tool choice:
Bash is the appropriate tool for this task. The script's purpose is Git tag management - fundamentally a command-line operation that requires executing git/gh commands, which bash excels at. The "non-standard libraries" you mention are:
- GitHub CLI (`gh`) - the official GitHub tool
- `jq` - industry-standard JSON processor
- Git itself

These are standard tools in our development environment and are explicitly documented in the requirements. Rewriting this in Python or another language would add unnecessary complexity, dependencies, and overhead for what is essentially orchestrating CLI commands.

Regarding maintainability:
The code is well-structured with descriptive function names, clear separation of concerns, and extensive comments. Anyone familiar with bash (a core skill for system administrators and DevOps engineers) can maintain this.

I welcome discussion about genuine improvements, but the suggestion that this is throwaway code or shouldn't be in the repository doesn't reflect the actual code quality or appropriateness of the implementation.

File: sbin/tagman [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+#   Git command-line tool: https://git-scm.com/
+# Warnings:
+# - This script modifies Git tags. Use with caution.
+# - Always verify the current tags before making changes.
+
+set -euo pipefail
+# Colour codes for output
+GRN='\033[0;32m'
+RED='\033[0;31m'
+YLW='\033[0;33m'
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: As above.

File: sbin/tagman [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+  $(basename "$0") add v1.0.0 abc123def --repo MetOffice/git_playground
+
+  # Create tag from existing tag
+  $(basename "$0") add Test vn1.5 --repo MetOffice/git_playground
+
+  # Create tag from release
+  $(basename "$0") add Autumn2025 vn14.0 --repo MetOffice/um
+
+  # Create tag from branch
+  $(basename "$0") add Spring2026 main --repo MetOffice/SimSys_Scripts
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: I understand the concern about screen real estate, but I respectfully disagree that this is excessive. Still I have a feeling that you can convince me to shorten this ;)

The usage function serves as both immediate help and documentation for users who may not have access to or won't read separate documentation. It provides:
- Clear syntax for all three actions (add/delete/list)
- Available options and their formats
- Five practical examples covering different use cases
- Important notes about behaviour

:warning: This script performs potentially destructive Git operations (creating/deleting tags in production repositories). Comprehensive examples aren't "overkill" - they help prevent user errors that could have significant consequences. Each example demonstrates a distinct use case:
1. Tag from commit SHA (most common)
1. Tag from existing tag
1. Tag from release (cross-repo workflow)
1. Tag from branch
1. Delete operation

Users can scroll or pipe to `less` if needed: `tagman | less`
Terminal heights vary, but scrolling is expected for help text
If there's a specific concern about this blocking adoption or causing issues, I'm open to discussing it, but cutting back examples would make the tool harder to use correctly, especially for destructive operations.

File: sbin/tagman [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+Notes:
+  - REPO can be set via environment variable (default: $DEFAULT_REPO)
+  - All other parameters must be provided via command-line arguments
+  - Use --dry-run to preview changes before executing
+EOF
+  exit 1
+}
+
+cleanup() {
+  [[ -n "${WORK_TMP:-}" ]] && rm -rf "$WORK_TMP"
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: I appreciate your attention to code clarity. The `[[ ]]` construct is actually the recommended modern bash syntax over `test` or `[ ]`. It's more robust, handles edge cases better (like empty strings), supports pattern matching, and is explicitly recommended in the Google Shell Style Guide and other authoritative sources. This isn't a "clever contraction" but rather best practice for bash scripting.

File: sbin/tagman [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+}
+
+confirm() {
+  local message="$1"
+  local response
+  echo -en "${YLW}"
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: As explained earlier.

File: sbin/tagman [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+  read -rp "$message (y/n): " response
+  echo -en "${NC}"
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: As above.

File: sbin/tagman [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+    local tag_info
+    if tag_info=$(gh api "repos/${REPO}/git/tags/${tag_sha}" 2>/dev/null); then
+      # It's an annotated tag - get the commit SHA it points to
+      local tag_type
+      tag_type=$(echo "$tag_info" | jq -r '.object.type')
+
+      if [[ "$tag_type" == "commit" ]]; then
+        resolved_sha=$(echo "$tag_info" | jq -r '.object.sha')
+      else
+        echo -e "${RED}** Tag points to unexpected object type: $tag_type${NC}"
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: Thank you for the suggestion. However, the current message is actually more accurate for future-proofing. While we currently only handle `commit` types, the Git object model supports multiple types (`blob`, `tree`, `commit`, `tag`), and the message correctly indicates that something unexpected was encountered. If we hardcode the message as suggested, it becomes less accurate if the code ever needs to handle other types.

File: sbin/tagman [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+    local branch_sha
+    branch_sha=$(gh api "repos/${REPO}/git/refs/heads/${REF}" --jq '.object.sha')
+
+    # Verify it's a full SHA
+    if resolved_sha=$(gh api "repos/${REPO}/commits/${branch_sha}" --jq '.sha' 2>/dev/null); then
+      REF="$resolved_sha"
+      echo -e "${GRN}Resolved to commit: $REF${NC}"
+      return 0
+    else
+      echo -e "${RED}** Failed to verify commit SHA from branch${NC}"
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: As above. Suggested change cannot be accepted.

File: sbin/tagman [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+      return 1
+    fi
+  fi
+
+  echo -e "${RED}** Reference '$REF' not found in repository '$REPO'.${NC}"
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: As above. Suggested change cannot be accepted.

File: sbin/tagman [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+  echo -e "${RED}** Tried: commit SHA, tag, release, and branch name.${NC}"
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: As above. Suggested change cannot be accepted.

File: sbin/tagman [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+  popd >/dev/null
+
+  echo -e "${GRN}Successfully created and pushed tag '$TAG' to '$REPO'.${NC}"
+}
+
+delete_tag() {
+  [[ -z "$TAG" ]] && {
+    echo -e "${RED}** TAG is required for delete action.${NC}"
+    usage
+  }
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: Suggested change cannot be applied.

File: sbin/tagman [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+
+parse_args() {
+  [[ $# -eq 0 ]] && usage
+
+  ACTION="$1"
+  shift
+
+  case "$ACTION" in
+    add)
+      [[ $# -lt 2 ]] && usage
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: As above.

File: sbin/tagman (hunk 4 above)
Comment (yaswant) [reply sbin/tagman]: Suggested change cannot be applied.

File: sbin/tagman [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+}
+
+main() {
+  parse_args "$@"
+
+  case "$ACTION" in
+    add) add_tag ;;
+    del|delete) delete_tag ;;
+    ls|list) list_tags ;;
+     *) echo -e "${RED}Invalid action: $ACTION${NC}" ;;
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: Suggested change cannot be applied.

File: sbin/tagman [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+# Requires:
+#   GitHub CLI (gh): https://cli.github.com/
+#   Git command-line tool: https://git-scm.com/
+# Warnings:
+# - This script modifies Git tags. Use with caution.
+# - Always verify the current tags before making changes.
+
+set -euo pipefail
+# Colour codes for output
+GRN='\033[0;32m'
```
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: I'm afraid I disagree.
The fact that I didn't know what NC was indicates it's not a sufficiently7 well known convention.

Variable names should be full, not abbreviated, words wherever possible for clarity.

File: sbin/tagman [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+RED='\033[0;31m'
+YLW='\033[0;33m'
+NC='\033[0m'
```
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: as above

File: sbin/tagman (hunk 15 above)
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: aas above

File: sbin/tagman (hunk 9 above)
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: as above

File: sbin/tagman [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+  return 1
+}
+
+verify_ref() {
+  local resolved_sha=""
+
+  # First, try to resolve as a commit SHA (handles both short and full)
+  if resolved_sha=$(gh api "repos/${REPO}/commits/${REF}" --jq '.sha' 2>/dev/null); then
+    REF="$resolved_sha"
+    echo -e "${GRN}Using commit SHA: $REF${NC}"
```
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: as above - I don't see why not.

File: sbin/tagman [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+    return 0
+  fi
+
+  # Try to resolve as a tag
+  if gh api "repos/${REPO}/git/refs/tags/${REF}" >/dev/null 2>&1; then
+    echo -e "${YLW}Resolving tag '$REF' to commit SHA...${NC}"
```
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: as above - I don't see why not.

File: sbin/tagman (hunk 10 above)
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: When the code changes, maker the comment more general purpose.
Until then make the comment reflect the code.

File: sbin/tagman [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+      resolved_sha="$tag_sha"
+    fi
+
+    # Verify it's a full SHA and a valid commit
+    if resolved_sha=$(gh api "repos/${REPO}/commits/${resolved_sha}" --jq '.sha' 2>/dev/null); then
+      REF="$resolved_sha"
+      echo -e "${GRN}Resolved to commit: $REF${NC}"
+      return 0
+    else
+      echo -e "${RED}** Failed to verify commit SHA from tag${NC}"
```
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: as above - I don't see why not.

File: sbin/tagman [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+      return 0
+    else
+      echo -e "${RED}** Failed to resolve release target to commit SHA${NC}"
+      return 1
+    fi
+  fi
+
+  # Try as a branch name
+  if gh api "repos/${REPO}/git/refs/heads/${REF}" >/dev/null 2>&1; then
+    echo -e "${YLW}Resolving branch '$REF' to commit SHA...${NC}"
```
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: as above - I don't see why not.

File: sbin/tagman (hunk 11 above)
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: as above - I don't see why not.

File: sbin/tagman [hunk 23]
```diff
@@ -0,0 +1,336 @@
...
+    local branch_sha
+    branch_sha=$(gh api "repos/${REPO}/git/refs/heads/${REF}" --jq '.object.sha')
+
+    # Verify it's a full SHA
+    if resolved_sha=$(gh api "repos/${REPO}/commits/${branch_sha}" --jq '.sha' 2>/dev/null); then
+      REF="$resolved_sha"
+      echo -e "${GRN}Resolved to commit: $REF${NC}"
```
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: as above - I don't see why not.

File: sbin/tagman [hunk 24]
```diff
@@ -0,0 +1,336 @@
...
+}
+
+add_tag() {
+  [[ -z "$TAG" || -z "$REF" ]] && {
+    echo -e "${RED}** TAG and REF are required for add action.${NC}"
+    usage
+  }
+
+  verify_tag && exit 1
+  verify_ref || exit 1
```
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: Please remove the "clever contraction" "||" as indicated

File: sbin/tagman [hunk 25]
```diff
@@ -0,0 +1,336 @@
...
+  pushd "$WORK_TMP" >/dev/null
+  run "Initialise temporary Git repository in $WORK_TMP" git init --bare --quiet
+  run "Add remote repository $url" git remote add origin "$url"
+  run "Fetch commit $REF" git fetch --quiet --depth 1 origin "$REF"
+  run "Create and sign tag '$TAG' at commit $REF" \
+    git tag --sign "$TAG" "$REF" --message "$msg"
+  run "Push '$TAG' to remote '$REPO'" git push --quiet origin "$TAG"
+  popd >/dev/null
+
+  echo -e "${GRN}Successfully created and pushed tag '$TAG' to '$REPO'.${NC}"
```
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: as above - I don't see why not.

File: sbin/tagman (hunk 15 above)
[the remark this answers is not in the document]
Comment (r-sharp) [reply sbin/tagman]: as above - I don't see why not.

File: sbin/tagman [hunk 26]
```diff
@@ -0,0 +1,336 @@
...
+  # Try as a branch name
+  if gh api "repos/${REPO}/git/refs/heads/${REF}" >/dev/null 2>&1; then
+    echo -e "${YLW}Resolving branch '$REF' to commit SHA...${NC}"
+    local branch_sha
+    branch_sha=$(gh api "repos/${REPO}/git/refs/heads/${REF}" --jq '.object.sha')
+
+    # Verify it's a full SHA
+    if resolved_sha=$(gh api "repos/${REPO}/commits/${branch_sha}" --jq '.sha' 2>/dev/null); then
+      REF="$resolved_sha"
+      echo -e "${GRN}Resolved to commit: $REF${NC}"
```
Comment (r-sharp) [inline sbin/tagman]:
```suggestion
      echo -e "${GREEN}Resolved to commit: $REF${NO_COLOUR}"
```

File: sbin/tagman [hunk 27]
```diff
@@ -0,0 +1,336 @@
...
+  local url="https://github.com/${REPO}.git"
+  local msg="${MESSAGE:-"Tagging $TAG @ $REF"}"
+
+  if [[ "$DRY_RUN" == true ]]; then
+    echo -e "${YLW}[DRY RUN] Would create tag with the following details:${NC}"
+    echo -e "  Repository: $REPO"
+    echo -e "  Tag name:   $TAG"
+    echo -e "  Commit SHA: $REF"
+    echo -e "  Message:    $msg"
+    echo -e "${YLW}[DRY RUN] No changes made.${NC}"
```
Comment (r-sharp) [inline sbin/tagman]:
```suggestion
    echo -e "${YELLOW}[DRY RUN] No changes made.${NO_COLOUR}"
```

File: sbin/tagman [hunk 28]
```diff
@@ -142,7 +142,7 @@ verify_ref() {
       if [[ "$tag_type" == "commit" ]]; then
         resolved_sha=$(echo "$tag_info" | jq -r '.object.sha')
       else
-        echo -e "${RED}** Tag points to unexpected object type: $tag_type${NC}"
+        echo -e "${RED}** Tag points to unexpected object type: $tag_type${NO_COLOR}"
```
[the remark this answers is not in the document]
Comment (yaswant) [reply sbin/tagman]: Done in https://github.com/MetOffice/SimSys_Scripts/pull/185/commits/8b5e10f0c78c28030db725860009cee0c2db2e1d
